<a href="https://colab.research.google.com/github/DeepakKumarSingh003/Book-Recommendation-System/blob/main/Book_Recommender_system.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import pandas as pd

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [3]:
df1=pd.read_csv("/content/drive/MyDrive/Recommender System/Book Recommendation System/Books.csv")
df2=pd.read_csv("/content/drive/MyDrive/Recommender System/Book Recommendation System/Ratings.csv")
df3=pd.read_csv("/content/drive/MyDrive/Recommender System/Book Recommendation System/Users.csv")

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2882: DtypeWarning: Columns (3) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


#Quick View at Data set

In [4]:
#Book Details
print(df1.shape)
df1.head()

(271360, 8)


ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [ ]:
df1.columns

Index(['ISBN', 'Book-Title', 'Book-Author', 'Year-Of-Publication', 'Publisher',
       'Image-URL-S', 'Image-URL-M', 'Image-URL-L'],
      dtype='object')

In [ ]:
df1['Image-URL-M'][1]

'http://images.amazon.com/images/P/0002005018.01.MZZZZZZZ.jpg'

In [ ]:
#Rating Details
print(df2.shape)
df2.head()

(1149780, 3)


User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

In [ ]:
df2['User-ID'].unique()
# df2["Book-Rating"].value_counts()

array([276725, 276726, 276727, ..., 276709, 276721, 276723])

In [ ]:
#User Details
df3.head()

User-ID                            Location   Age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN

#Check Nulls and Duplocates

**Nulls**

In [ ]:

print("************************Book******************************")
print(df1.isnull().sum())
print("***********************Ratings*****************************")
print(df2.isnull().sum())
print("************************Users*******************************")
print(df3.isnull().sum())

************************Book******************************
ISBN                   0
Book-Title             0
Book-Author            1
Year-Of-Publication    0
Publisher              2
Image-URL-S            0
Image-URL-M            0
Image-URL-L            3
dtype: int64
***********************Ratings*****************************
User-ID        0
ISBN           0
Book-Rating    0
dtype: int64
************************Users*******************************
User-ID          0
Location         0
Age         110762
dtype: int64


In [ ]:
# df1=df1.dropna(inplace=True)

**Duplicates**

In [ ]:
print("************************Book******************************")
print(df1.duplicated().sum())
print("***********************Ratings*****************************")
print(df2.duplicated().sum())
print("************************Users*******************************")
print(df3.duplicated().sum())

************************Book******************************
0
***********************Ratings*****************************
0
************************Users*******************************
0


#Popularity Based Recommender

* **Merge the books and rating data frame on isbn**
* **

In [ ]:
merge_df=df1.merge(df2,on="ISBN")
merge_df.head(2)

ISBN           Book-Title           Book-Author Year-Of-Publication  \
0  0195153448  Classical Mythology    Mark P. O. Morford                2002   
1  0002005018         Clara Callan  Richard Bruce Wright                2001   

                 Publisher                                        Image-URL-S  \
0  Oxford University Press  http://images.amazon.com/images/P/0195153448.0...   
1    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-L  User-ID  Book-Rating  
0  http://images.amazon.com/images/P/0195153448.0...        2            0  
1  http://images.amazon.com/images/P/0002005018.0...        8            5

In [ ]:
rating=merge_df.groupby("Book-Title").count()["Book-Rating"].reset_index()
rating.rename(columns={"Book-Rating":"Num_of_Rating"},inplace=True)
rating

Book-Title  Num_of_Rating
0        A Light in the Storm: The Civil War Diary of ...              4
1                                   Always Have Popsicles              1
2                    Apple Magic (The Collector's series)              1
3        Ask Lily (Young Women of Faith: Lily Series, ...              1
4        Beyond IBM: Leadership Marketing and Finance ...              1
...                                                   ...            ...
241066                                      Ã?Â?lpiraten.              2
241067                     Ã?Â?rger mit Produkt X. Roman.              4
241068                                Ã?Â?sterlich leben.              1
241069                              Ã?Â?stlich der Berge.              3
241070                                  Ã?Â?thique en toc              2

[241071 rows x 2 columns]

In [ ]:
avg_rating=merge_df.groupby("Book-Title").mean()["Book-Rating"].reset_index()
avg_rating.rename(columns={"Book-Rating":"avg_of_Rating"},inplace=True)
avg_rating

Book-Title  avg_of_Rating
0        A Light in the Storm: The Civil War Diary of ...       2.250000
1                                   Always Have Popsicles       0.000000
2                    Apple Magic (The Collector's series)       0.000000
3        Ask Lily (Young Women of Faith: Lily Series, ...       8.000000
4        Beyond IBM: Leadership Marketing and Finance ...       0.000000
...                                                   ...            ...
241066                                      Ã?Â?lpiraten.       0.000000
241067                     Ã?Â?rger mit Produkt X. Roman.       5.250000
241068                                Ã?Â?sterlich leben.       7.000000
241069                              Ã?Â?stlich der Berge.       2.666667
241070                                  Ã?Â?thique en toc       4.000000

[241071 rows x 2 columns]

In [ ]:
popular_df= rating.merge(avg_rating,on="Book-Title")
popular_df

Book-Title  Num_of_Rating  \
0        A Light in the Storm: The Civil War Diary of ...              4   
1                                   Always Have Popsicles              1   
2                    Apple Magic (The Collector's series)              1   
3        Ask Lily (Young Women of Faith: Lily Series, ...              1   
4        Beyond IBM: Leadership Marketing and Finance ...              1   
...                                                   ...            ...   
241066                                      Ã?Â?lpiraten.              2   
241067                     Ã?Â?rger mit Produkt X. Roman.              4   
241068                                Ã?Â?sterlich leben.              1   
241069                              Ã?Â?stlich der Berge.              3   
241070                                  Ã?Â?thique en toc              2   

        avg_of_Rating  
0            2.250000  
1            0.000000  
2            0.000000  
3            8.000000  
4            0.000000  
...               ...  
241066       0.000000  
241067       5.250000  
241068       7.000000  
241069       2.666667  
241070       4.000000  

[241071 rows x 3 columns]

In [ ]:
popular_df=popular_df[popular_df["Num_of_Rating"] >= 250].sort_values("avg_of_Rating",ascending=False).head(50)
popular_df

Book-Title  Num_of_Rating  \
80434   Harry Potter and the Prisoner of Azkaban (Book 3)            428   
80422        Harry Potter and the Goblet of Fire (Book 4)            387   
80441      Harry Potter and the Sorcerer's Stone (Book 1)            278   
80426   Harry Potter and the Order of the Phoenix (Boo...            347   
80414    Harry Potter and the Chamber of Secrets (Book 2)            556   
191612  The Hobbit : The Enchanting Prelude to The Lor...            281   
187377  The Fellowship of the Ring (The Lord of the Ri...            368   
80445   Harry Potter and the Sorcerer's Stone (Harry P...            575   
211384     The Two Towers (The Lord of the Rings, Part 2)            260   
219741                              To Kill a Mockingbird            510   
183573                                  The Da Vinci Code            898   
187880                 The Five People You Meet in Heaven            430   
180556                             The Catcher in the Rye            449   
196326                          The Lovely Bones: A Novel           1295   
764                                                  1984            284   
144165                           Prodigal Summer: A Novel            253   
128670                                         Neverwhere            265   
206502                            The Secret Life of Bees            774   
168719  Stupid White Men ...and Other Sorry Excuses fo...            283   
223135  Tuesdays with Morrie: An Old Man, a Young Man,...            493   
204387                The Red Tent (Bestselling Backlist)            723   
191589               The Hitchhiker's Guide to the Galaxy            268   
129379   Nickel and Dimed: On (Not) Getting By in America            335   
93381                                       Into the Wild            252   
63867                                      Fahrenheit 451            409   
74750                           Girl with a Pearl Earring            526   
136145                                          Outlander            283   
233370  Where the Heart Is (Oprah's Book Club (Paperba...            585   
156102                     Seabiscuit: An American Legend            275   
107962                                         Life of Pi            664   
176845                                     The Bean Trees            389   
2281    A Child Called \It\": One Child's Courage to S...            265   
8434                                       ANGELA'S ASHES            279   
76343                                         Good in Bed            490   
64931   Fast Food Nation: The Dark Side of the All-Ame...            321   
12700                                       American Gods            302   
161645                                 Skipping Christmas            322   
105777  Left Behind: A Novel of the Earth's Last Days ...            318   
189551    The Golden Compass (His Dark Materials, Book 1)            336   
181679                                   The Color Purple            314   
160336                               Silence of the Lambs            256   
8752                                          About a Boy            262   
158138                  Seven Up (A Stephanie Plum Novel)            278   
175097  The Alchemist: A Fable About Following Your Dream            266   
80069   Hard Eight : A Stephanie Plum Novel (A Stephan...            269   
170101                       Suzanne's Diary for Nicholas            457   
111073                                  Lord of the Flies            259   
5664                              A Prayer for Owen Meany            413   
212070   The Vampire Lestat (Vampire Chronicles, Book II)            301   
233851       White Oleander : A Novel (Oprah's Book Club)            356   

        avg_of_Rating  
80434        5.852804  
80422        5.824289  
80441        5.737410  
80426        5.501441  
80414        5.183453  
191612       5.007117  
18

In [ ]:
display(df1.head(2),popular_df.head(2))

ISBN           Book-Title           Book-Author Year-Of-Publication  \
0  0195153448  Classical Mythology    Mark P. O. Morford                2002   
1  0002005018         Clara Callan  Richard Bruce Wright                2001   

                 Publisher                                        Image-URL-S  \
0  Oxford University Press  http://images.amazon.com/images/P/0195153448.0...   
1    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...

Book-Title  Num_of_Rating  \
80434  Harry Potter and the Prisoner of Azkaban (Book 3)            428   
80422       Harry Potter and the Goblet of Fire (Book 4)            387   

       avg_of_Rating  
80434       5.852804  
80422       5.824289

**Popular Books**

**Top 50 Books from the data set**

In [ ]:
popular_df=popular_df.merge(df1,on="Book-Title").drop_duplicates('Book-Title')[['Book-Title', 'Book-Author','Num_of_Rating', 'avg_of_Rating','Image-URL-M']]
popular_df

Book-Title  \
0    Harry Potter and the Prisoner of Azkaban (Book 3)   
3         Harry Potter and the Goblet of Fire (Book 4)   
5       Harry Potter and the Sorcerer's Stone (Book 1)   
9    Harry Potter and the Order of the Phoenix (Boo...   
13    Harry Potter and the Chamber of Secrets (Book 2)   
16   The Hobbit : The Enchanting Prelude to The Lor...   
17   The Fellowship of the Ring (The Lord of the Ri...   
26   Harry Potter and the Sorcerer's Stone (Harry P...   
28      The Two Towers (The Lord of the Rings, Part 2)   
39                               To Kill a Mockingbird   
47                                   The Da Vinci Code   
53                  The Five People You Meet in Heaven   
55                              The Catcher in the Rye   
62                           The Lovely Bones: A Novel   
63                                                1984   
72                            Prodigal Summer: A Novel   
73                                          Neverwhere   
78                             The Secret Life of Bees   
84   Stupid White Men ...and Other Sorry Excuses fo...   
85   Tuesdays with Morrie: An Old Man, a Young Man,...   
88                 The Red Tent (Bestselling Backlist)   
89                The Hitchhiker's Guide to the Galaxy   
98    Nickel and Dimed: On (Not) Getting By in America   
100                                      Into the Wild   
103                                     Fahrenheit 451   
116                          Girl with a Pearl Earring   
117                                          Outlander   
123  Where the Heart Is (Oprah's Book Club (Paperba...   
124                     Seabiscuit: An American Legend   
126                                         Life of Pi   
130                                     The Bean Trees   
133  A Child Called \It\": One Child's Courage to S...   
135                                     ANGELA'S ASHES   
137                                        Good in Bed   
139  Fast Food Nation: The Dark Side of the All-Ame...   
140                                      American Gods   
141                                 Skipping Christmas   
147  Left Behind: A Novel of the Earth's Last Days ...   
149    The Golden Compass (His Dark Materials, Book 1)   
157                                   The Color Purple   
163                               Silence of the Lambs   
166                                        About a Boy   
173                  Seven Up (A Stephanie Plum Novel)   
175  The Alchemist: A Fable About Following Your Dream   
177  Hard Eight : A Stephanie Plum Novel (A Stephan...   
182                       Suzanne's Diary for Nicholas   
186                                  Lord of the Flies   
192                            A Prayer for Owen Meany   
194   The Vampire Lestat (Vampire Chronicles, Book II)   
195       White Oleander : A Novel (Oprah's Book Club)   

                Book-Author  Num_of_Rating  avg_of_Rating  \
0             J. K. Rowling            428       5.852804   
3             J. K. Rowling            387       5.824289   
5             J. K. Rowling            278       5.737410   
9             J. K. Rowling            347       5.501441   
13            J. K. Rowling            556       5.183453   
16           J.R.R. TOLKIEN            281       5.007117   
17           J.R.R. TOLKIEN            368       4.948370   
26            J. K. Rowling            575       4.895652   
28           J.R.R. TOLKIEN            260       4.880769   
39               Harper Lee            510       4.700000   
47                Dan Brown            898       4.642539   
53              Mitch Albom            430       4.551163   
55            J.D. Salinger            449       4.545657   
62             Alice Sebold           1295       4.468726   
63            George Orwell            284       4.454225   
72       Barbara Kingsolver            253       4.450593   
73              Neil Gaiman            265      

#Collaborative Filterning Based Recommender System

* **Filter the users and Book-Title on Book_Ratings**

###Filter Only those user who have Rated the Books more than 200 times.

In [ ]:
merge_df.head()

ISBN           Book-Title           Book-Author Year-Of-Publication  \
0  0195153448  Classical Mythology    Mark P. O. Morford                2002   
1  0002005018         Clara Callan  Richard Bruce Wright                2001   
2  0002005018         Clara Callan  Richard Bruce Wright                2001   
3  0002005018         Clara Callan  Richard Bruce Wright                2001   
4  0002005018         Clara Callan  Richard Bruce Wright                2001   

                 Publisher                                        Image-URL-S  \
0  Oxford University Press  http://images.amazon.com/images/P/0195153448.0...   
1    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   
2    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   
3    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   
4    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0002005018.0...   
3  http://images.amazon.com/images/P/0002005018.0...   
4  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-L  User-ID  Book-Rating  
0  http://images.amazon.com/images/P/0195153448.0...        2            0  
1  http://images.amazon.com/images/P/0002005018.0...        8            5  
2  http://images.amazon.com/images/P/0002005018.0...    11400            0  
3  http://images.amazon.com/images/P/0002005018.0...    11676            8  
4  http://images.amazon.com/images/P/0002005018.0...    41385            0

In [ ]:
#Merged df of Books and Rating
merge_df.head(2)
print(merge_df["User-ID"].value_counts())
Users_who_Rated=merge_df.groupby("User-ID").count()["Book-Rating"]
Users_who_Rated.reset_index()

11676     11144
198711     6456
153662     5814
98391      5779
35859      5646
          ...  
51911         1
151772        1
158802        1
243672        1
259920        1
Name: User-ID, Length: 92106, dtype: int64


User-ID  Book-Rating
0            2            1
1            8           17
2            9            3
3           10            1
4           12            1
...        ...          ...
92101   278846            1
92102   278849            4
92103   278851           23
92104   278852            1
92105   278854            8

[92106 rows x 2 columns]

**Observation**
* **Out of 259920 Users Only 92106 Users Rated the Books.**

###All Those Users who have minimum ratinged 200 times.

In [ ]:
x=merge_df.groupby("User-ID").count()["Book-Rating"] >= 200
Padhe_likhe_users=x[x].index
Padhe_likhe_users

Int64Index([   254,   2276,   2766,   2977,   3363,   4017,   4385,   6251,
              6323,   6543,
            ...
            271705, 273979, 274004, 274061, 274301, 274308, 275970, 277427,
            277639, 278418],
           dtype='int64', name='User-ID', length=811)

In [ ]:
Filter_Ratings=merge_df[merge_df["User-ID"].isin(Padhe_likhe_users)]
print(Filter_Ratings.shape)
Filter_Ratings.head()

(474007, 10)


ISBN                                         Book-Title  \
3   0002005018                                       Clara Callan   
6   0002005018                                       Clara Callan   
7   0002005018                                       Clara Callan   
10  0002005018                                       Clara Callan   
21  0374157065  Flu: The Story of the Great Influenza Pandemic...   

             Book-Author Year-Of-Publication              Publisher  \
3   Richard Bruce Wright                2001  HarperFlamingo Canada   
6   Richard Bruce Wright                2001  HarperFlamingo Canada   
7   Richard Bruce Wright                2001  HarperFlamingo Canada   
10  Richard Bruce Wright                2001  HarperFlamingo Canada   
21      Gina Bari Kolata                1999   Farrar Straus Giroux   

                                          Image-URL-S  \
3   http://images.amazon.com/images/P/0002005018.0...   
6   http://images.amazon.com/images/P/0002005018.0...   
7   http://images.amazon.com/images/P/0002005018.0...   
10  http://images.amazon.com/images/P/0002005018.0...   
21  http://images.amazon.com/images/P/0374157065.0...   

                                          Image-URL-M  \
3   http://images.amazon.com/images/P/0002005018.0...   
6   http://images.amazon.com/images/P/0002005018.0...   
7   http://images.amazon.com/images/P/0002005018.0...   
10  http://images.amazon.com/images/P/0002005018.0...   
21  http://images.amazon.com/images/P/0374157065.0...   

                                          Image-URL-L  User-ID  Book-Rating  
3   http://images.amazon.com/images/P/0002005018.0...    11676            8  
6   http://images.amazon.com/images/P/0002005018.0...    85526            0  
7   http://images.amazon.com/images/P/0002005018.0...    96054            0  
10  http://images.amazon.com/images/P/0002005018.0...   177458            0  
21  http://images.amazon.com/images/P/0374157065.0...   110912           10

###Filte Book By number of Ratings.

In [ ]:
y=Filter_Ratings.groupby("Book-Title").count()["Book-Rating"]>=50
Famous_books=y[y].index
Famous_books

Index(['1984', '1st to Die: A Novel', '2nd Chance', '4 Blondes',
       'A Bend in the Road', 'A Case of Need',
       'A Child Called \It\": One Child's Courage to Survive"',
       'A Civil Action', 'A Day Late and a Dollar Short', 'A Fine Balance',
       ...
       'Winter Solstice', 'Wish You Well', 'Without Remorse',
       'Wizard and Glass (The Dark Tower, Book 4)', 'Wuthering Heights',
       'Year of Wonders', 'You Belong To Me',
       'Zen and the Art of Motorcycle Maintenance: An Inquiry into Values',
       'Zoya', '\O\" Is for Outlaw"'],
      dtype='object', name='Book-Title', length=706)

In [ ]:
final_rating=Filter_Ratings[Filter_Ratings['Book-Title'].isin(Famous_books)]
print(final_rating.duplicated().sum())
print(final_rating.shape)
final_rating


0
(58586, 10)


ISBN                       Book-Title      Book-Author  \
31       0399135782           The Kitchen God's Wife          Amy Tan   
33       0399135782           The Kitchen God's Wife          Amy Tan   
34       0399135782           The Kitchen God's Wife          Amy Tan   
38       0399135782           The Kitchen God's Wife          Amy Tan   
39       0399135782           The Kitchen God's Wife          Amy Tan   
...             ...                              ...              ...   
1028414  1878702831                           Echoes      Nancy Morse   
1028600  0394429869  I Know Why the Caged Bird Sings     Maya Angelou   
1028602  0449001164                      The Promise      CHAIM POTOK   
1028815  0743527631         The Pillars of the Earth      Ken Follett   
1028817  0745168086              The Handmaid's Tale  Margaret Atwood   

        Year-Of-Publication                      Publisher  \
31                     1991               Putnam Pub Group   
33                     1991               Putnam Pub Group   
34                     1991               Putnam Pub Group   
38                     1991               Putnam Pub Group   
39                     1991               Putnam Pub Group   
...                     ...                            ...   
1028414                1992  Meteor Publishing Corporation   
1028600                1996                   Random House   
1028602                1997               Ballantine Books   
1028815                2002                         Encore   
1028817                1999            Chivers Audio Books   

                                               Image-URL-S  \
31       http://images.amazon.com/images/P/0399135782.0...   
33       http://images.amazon.com/images/P/0399135782.0...   
34       http://images.amazon.com/images/P/0399135782.0...   
38       http://images.amazon.com/images/P/0399135782.0...   
39       http://images.amazon.com/images/P/0399135782.0...   
...                                                    ...   
1028414  http://images.amazon.com/images/P/1878702831.0...   
1028600  http://images.amazon.com/images/P/0394429869.0...   
1028602  http://images.amazon.com/images/P/0449001164.0...   
1028815  http://images.amazon.com/images/P/0743527631.0...   
1028817  http://images.amazon.com/images/P/0745168086.0...   

                                               Image-URL-M  \
31       http://images.amazon.com/images/P/0399135782.0...   
33       http://images.amazon.com/images/P/0399135782.0...   
34       http://images.amazon.com/images/P/0399135782.0...   
38       http://images.amazon.com/images/P/0399135782.0...   
39       http://images.amazon.com/images/P/0399135782.0...   
...                                                    ...   
1028414  http://images.amazon.com/images/P/1878702831.0...   
1028600  http://images.amazon.com/images/P/0394429869.0...   
1028602  http://images.amazon.com/images/P/0449001164.0...   
1028815  http://images.amazon.com/images/P/0743527631.0...   
1028817  http://images.amazon.com/images/P/0745168086.0...   

                                               Image-URL-L  User-ID  \
31       http://images.amazon.com/images/P/0399135782.0...    11676   
33       http://images.amazon.com/images/P/0399135782.0...    36836   
34       http://images.amazon.com/images/P/0399135782.0...    46398   
38       http://images.amazon.com/images/P/0399135782.0...   113270   
39       http://images.amazon.com/images/P/0399135782.0...   113519   
...                                                    ...      ...   
1028414  http://images.amazon.com/images/P/1878702831.0...   238781   
1028600  http://images.amazon.com/images/P/0394429869.0...   239594   
1028602  http://images.amazon.com/images/P/0449001164.0...   239594   
1028815  http://images.amazon.com/images/P/0743527631.0...   240144   
1028817  http://images.amazon.com/images/P/0745168086.0...   240144   

         Book-Rating  
31                 9  
33  

###Pivot table

In [ ]:
Pivot_Table=final_rating.pivot_table(index="Book-Title",columns="User-ID",values="Book-Rating")
Pivot_Table

User-ID                                             254     2276    2766    \
Book-Title                                                                   
1984                                                   9.0     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN    10.0     NaN   
4 Blondes                                              NaN     NaN     NaN   
A Bend in the Road                                     0.0     NaN     7.0   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     NaN     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             2977    3363    4017    \
Book-Title                                                                   
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     NaN     NaN   
A Bend in the Road                                     NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     0.0     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             4385    6251    6323    \
Book-Title                                                                   
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     0.0     NaN   
A Bend in the Road                                     NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     NaN   
You Belong To Me                                       NaN     NaN     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     0.0     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             6543    ...  271705  \
Book-Title                                                  ...           
1984                                                   NaN  ...    10.0   
1st to Die: A Novel                                    9.0  ...     NaN   
2nd Chance                                             0.0  ...     NaN   
4 Blondes                                              NaN  ...     NaN   
A Bend in the Road                                     NaN  ...     NaN   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     NaN   
You Belong To Me                                       NaN  ...     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN  ...     NaN   
Zoya                                                   NaN  ...     NaN   
\O\" Is for Outlaw"                                    

In [ ]:
#Fill Nan with 0
Pivot_Table.fillna(0,inplace=True)
Pivot_Table

User-ID                                             254     2276    2766    \
Book-Title                                                                   
1984                                                   9.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0    10.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     7.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             2977    3363    4017    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             4385    6251    6323    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             6543    ...  271705  \
Book-Title                                                  ...           
1984                                                   0.0  ...    10.0   
1st to Die: A Novel                                    9.0  ...     0.0   
2nd Chance                                             0.0  ...     0.0   
4 Blondes                                              0.0  ...     0.0   
A Bend in the Road                                     0.0  ...     0.0   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     0.0   
You Belong To Me                                       0.0  ...     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     0.0   
Zoya                                                   0.0  ...     0.0   
\O\" Is for Outlaw"                                    

**Observation**
* **In the final ratings we 706 famous books which are been rated more than 50 times**.
* **We have 810 User who have Rated more than 200 times**. 

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity
similarity_score=cosine_similarity(Pivot_Table)
similarity_score

array([[1.        , 0.10255025, 0.01220856, ..., 0.12110367, 0.07347567,
        0.04316046],
       [0.10255025, 1.        , 0.2364573 , ..., 0.07446129, 0.16773875,
        0.14263397],
       [0.01220856, 0.2364573 , 1.        , ..., 0.04558758, 0.04938579,
        0.10796119],
       ...,
       [0.12110367, 0.07446129, 0.04558758, ..., 1.        , 0.07085128,
        0.0196177 ],
       [0.07347567, 0.16773875, 0.04938579, ..., 0.07085128, 1.        ,
        0.10602962],
       [0.04316046, 0.14263397, 0.10796119, ..., 0.0196177 , 0.10602962,
        1.        ]])

In [ ]:
#Book 1 cosine score with all the other Books
print(similarity_score.shape)
similarity_score[0]

(706, 706)


706

##Recommend a Book

In [ ]:
#From the pivote table take the index value of the book
np.where(Pivot_Table.index=="1984")
np.where(Pivot_Table.index=="1st to Die: A Novel")[0]#[0]
np.where(Pivot_Table.index=='\O\" Is for Outlaw')[0][0]

(array([], dtype=int64),)

In [ ]:
similarity_score[0]

array([1.        , 0.10255025, 0.01220856, 0.        , 0.05367224,
       0.02774901, 0.08216491, 0.13732869, 0.03261686, 0.03667591,
       0.02322418, 0.06766487, 0.02083978, 0.09673735, 0.13388865,
       0.08303112, 0.11153543, 0.05100411, 0.02517784, 0.11706383,
       0.        , 0.14333793, 0.07847534, 0.06150451, 0.08723968,
       0.        , 0.07009814, 0.13658681, 0.07600328, 0.12167134,
       0.00768046, 0.01473221, 0.        , 0.07965814, 0.04522617,
       0.01556271, 0.09495938, 0.0182307 , 0.02610465, 0.07984012,
       0.11679969, 0.0569124 , 0.08354155, 0.08471898, 0.08785938,
       0.05491435, 0.0548505 , 0.27026514, 0.09779123, 0.06016046,
       0.08958835, 0.06748675, 0.        , 0.04468098, 0.01920872,
       0.        , 0.05629067, 0.00557964, 0.07877059, 0.05219479,
       0.18908177, 0.        , 0.01240656, 0.02984572, 0.04279502,
       0.12680125, 0.16566735, 0.        , 0.13357242, 0.06615478,
       0.        , 0.        , 0.        , 0.10968075, 0.02806

In [ ]:
x=list(enumerate(similarity_score[1]))
x[1:6]

[(1, 1.0),
 (2, 0.23645729525444303),
 (3, 0.0),
 (4, 0.10953789733294741),
 (5, 0.10092909036815481)]

In [ ]:
#Take the index and check the similarity score of that book with other books and select only 5 most similar books
similarity_score[1] #Select book
enumerate(similarity_score[1]) #have the position
list(enumerate(similarity_score[1])) #store position and similarity_score
sorted(list(enumerate(similarity_score[1])),key=lambda x:x[1],reverse=True) #sort by similarity score
sorted(list(enumerate(similarity_score[1])),key=lambda x:x[1],reverse=True)[1:6] #select top 6 rejecting self(i;e fro 1 to 6)

[(39, 0.30746647647493613),
 (407, 0.2827397366387593),
 (382, 0.2780357116802063),
 (666, 0.2776809645250099),
 (303, 0.2715497227877098)]

In [ ]:
Pivot_Table.index

Index(['1984', '1st to Die: A Novel', '2nd Chance', '4 Blondes',
       'A Bend in the Road', 'A Case of Need',
       'A Child Called \It\": One Child's Courage to Survive"',
       'A Civil Action', 'A Day Late and a Dollar Short', 'A Fine Balance',
       ...
       'Winter Solstice', 'Wish You Well', 'Without Remorse',
       'Wizard and Glass (The Dark Tower, Book 4)', 'Wuthering Heights',
       'Year of Wonders', 'You Belong To Me',
       'Zen and the Art of Motorcycle Maintenance: An Inquiry into Values',
       'Zoya', '\O\" Is for Outlaw"'],
      dtype='object', name='Book-Title', length=706)

In [ ]:
df1.head(2)

ISBN           Book-Title           Book-Author Year-Of-Publication  \
0  0195153448  Classical Mythology    Mark P. O. Morford                2002   
1  0002005018         Clara Callan  Richard Bruce Wright                2001   

                 Publisher                                        Image-URL-S  \
0  Oxford University Press  http://images.amazon.com/images/P/0195153448.0...   
1    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...

In [ ]:
index=np.where(Pivot_Table.index=="1984")[0][0]
similarity=sorted(list(enumerate(similarity_score[index])),key=lambda x:x[1],reverse=True)[1:6] 
print(similarity)

[(47, 0.2702651417103732), (545, 0.2639619371123496), (82, 0.2366937434740099), (634, 0.23299389358170397), (551, 0.2262639743141286)]


In [ ]:
for i in similarity:
  print(Pivot_Table.index[i[0]])

Animal Farm
The Handmaid's Tale
Brave New World
The Vampire Lestat (Vampire Chronicles, Book II)
The Hours : A Novel


In [ ]:
def recommend(Book):
  index=np.where(Pivot_Table.index==Book)[0][0]
  similarity=sorted(list(enumerate(similarity_score[index])),key=lambda x:x[1],reverse=True)[1:6] 
  print(similarity)
  
  data=[]
  for i in similarity:
    item=[]
    temp_df=df1[df1['Book-Title'] == Pivot_Table.index[i[0]]]
    item.extend(list(temp_df.drop_duplicates('Book-Title')["Book-Title"].values))
    item.extend(list(temp_df.drop_duplicates('Book-Title')["Book-Author"].values))
    item.extend(list(temp_df.drop_duplicates('Book-Title')["Image-URL-M"].values))
                
    data.append(item)
  return data

  # return suggestion

In [ ]:
#Test care
print(recommend("1984"))
# print("*"*10,"Year of Wonders","*"*10)
# recommend("Year of Wonders")

[(47, 0.2702651417103732), (545, 0.2639619371123496), (82, 0.2366937434740099), (634, 0.23299389358170397), (551, 0.2262639743141286)]
[['Animal Farm', 'George Orwell', 'http://images.amazon.com/images/P/0451526341.01.MZZZZZZZ.jpg'], ["The Handmaid's Tale", 'Margaret Atwood', 'http://images.amazon.com/images/P/0449212602.01.MZZZZZZZ.jpg'], ['Brave New World', 'Aldous Huxley', 'http://images.amazon.com/images/P/0060809833.01.MZZZZZZZ.jpg'], ['The Vampire Lestat (Vampire Chronicles, Book II)', 'ANNE RICE', 'http://images.amazon.com/images/P/0345313860.01.MZZZZZZZ.jpg'], ['The Hours : A Novel', 'Michael Cunningham', 'http://images.amazon.com/images/P/0312243022.01.MZZZZZZZ.jpg']]


#Pickle

####Popular based recommendatio

In [ ]:
import pickle
pickle.dump(popular_df,open('popular.pkl','wb'))

####Collaborative based recommendation

In [ ]:
import pickle
pickle.dump(Pivot_Table, open('Pivote_Table.pkl',"wb"))
pickle.dump(df1, open('books.pkl',"wb"))
pickle.dump(similarity_score, open('similarity_score.pkl',"wb"))